File: rd-text/src/main/kotlin/com/jetbrains/rider/rdtext/impl/ot/intrinsics/RdAck.kt
package com.jetbrains.rider.rdtext.impl.ot.intrinsics

import com.jetbrains.rider.framework.*
import com.jetbrains.rider.util.string.IPrintable
import com.jetbrains.rider.util.string.PrettyPrinter
import com.jetbrains.rider.util.string.print
import kotlin.reflect.KClass

data class RdAck (
        val timestamp : Int,
        val origin : com.jetbrains.rider.rdtext.impl.intrinsics.RdChangeOrigin
) : IPrintable {
    //companion

    companion object : IMarshaller<RdAck> {
        override val _type: KClass<RdAck> = RdAck::class

        @Suppress("UNCHECKED_CAST")
        override fun read(ctx: SerializationCtx, buffer: AbstractBuffer): RdAck {
            val timestamp = buffer.readInt()
            val origin = buffer.readEnum<com.jetbrains.rider.rdtext.impl.intrinsics.RdChangeOrigin>()
            return RdAck(timestamp, origin)
        }

        override fun write(ctx: SerializationCtx, buffer: AbstractBuffer, value: RdAck) {
            buffer.writeInt(value.timestamp)
            buffer.writeEnum(value.origin)
        }

    }
    //fields
    //initializer
    //secondary constructor
    //equals trait
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other?.javaClass != javaClass) return false

        other as RdAck

        if (timestamp != other.timestamp) return false
        if (origin != other.origin) return false

        return true
    }
    //hash code trait
    override fun hashCode(): Int {
        var __r = 0
        __r = __r*31 + timestamp.hashCode()
        __r = __r*31 + origin.hashCode()
        return __r
    }
    //pretty print
    override fun print(printer: PrettyPrinter) {
        printer.println("RdAck (")
        printer.indent {
            print("timestamp = "); timestamp.print(printer); println()
            print("origin = "); origin.print(printer); println()
        }
        printer.print(")")
    }
}

File: rd-net/RdFramework/Text/Impl/Ot/Intrinsics/RdAck.cs
using System;
using JetBrains.Rd.Base;
using JetBrains.Rd.Text.Impl.Intrinsics;
using JetBrains.Rd.Util;

namespace JetBrains.Rd.Text.Impl.Ot.Intrinsics
{
  public class RdAck : IPrintable, IEquatable<RdAck> {
    //fields
    //public fields
    public int Timestamp {get; private set;}
    public RdChangeOrigin Origin {get; private set;}

    //private fields
    //primary constructor
    public RdAck(
      int timestamp,
      RdChangeOrigin origin
    )
    {
      Timestamp = timestamp;
      Origin = origin;
    }
    //secondary constructor
    //statics

    public static CtxReadDelegate<RdAck> Read = (ctx, reader) =>
    {
      var timestamp = reader.ReadInt();
      var origin = (RdChangeOrigin)reader.ReadInt();
      return new RdAck(timestamp, origin);
    };

    public static CtxWriteDelegate<RdAck> Write = (ctx, writer, value) =>
    {
      writer.Write(value.Timestamp);
      writer.Write((int)value.Origin);
    };
    //custom body
    //equals trait
    public override bool Equals(object obj)
    {
      if (ReferenceEquals(null, obj)) return false;
      if (ReferenceEquals(this, obj)) return true;
      if (obj.GetType() != GetType()) return false;
      return Equals((RdAck) obj);
    }
    public bool Equals(RdAck other)
    {
      if (ReferenceEquals(null, other)) return false;
      if (ReferenceEquals(this, other)) return true;
      return Timestamp == other.Timestamp && Origin == other.Origin;
    }
    //hash code trait
    public override int GetHashCode()
    {
      unchecked {
        var hash = 0;
        hash = hash * 31 + Timestamp.GetHashCode();
        hash = hash * 31 + (int) Origin;
        return hash;
      }
    }
    //pretty print
    public void Print(PrettyPrinter printer)
    {
      printer.Println("RdAck (");
      using (printer.IndentCookie()) {
        printer.Print("timestamp = "); Timestamp.PrintEx(printer); printer.Println();
        printer.Print("origin = "); Origin.PrintEx(printer); printer.Println();
      }
      printer.Print(")");
    }
    //toString
    public override string ToString()
    {
      var printer = new SingleLinePrettyPrinter();
      Print(printer);
      return printer.ToString();
    }
  }
}